File: LeetCode/Easy/0020-valid-parentheses/0020-valid-parentheses.cpp
#include <string>
#include <stack>

class Solution {
public:
    /**
     * Determines if the input string of brackets is valid without a map template.
     * * INTUITION:
     * Utilizing sequential `if-else` branches to place anticipated closing brackets 
     * onto the stack maximizes runtime efficiency in compiled environments by 
     * avoiding hashing computations.
     * * COMPLEXITY ANALYSIS:
     * - Time Complexity: O(n)
     * - Space Complexity: O(n)
     */
    bool isValid(std::string s) {
        // Early exit optimization
        if (s.length() % 2 != 0) {
            return false;
        }

        std::stack<char> bracket_stack;

        for (char ch : s) {
            if (ch == '(') {
                bracket_stack.push(')');
            } else if (ch == '{') {
                bracket_stack.push('}');
            } else if (ch == '[') {
                bracket_stack.push(']');
            } 
            else {
                if (bracket_stack.empty() || bracket_stack.top() != ch) {
                    return false;
                }
                bracket_stack.pop();
            }
        }

        return bracket_stack.empty();
    }
};

File: LeetCode/Easy/0020-valid-parentheses/0020-valid-parentheses.java
import java.util.Stack;

class Solution {
    /**
     * Determines if the input string of brackets is valid without a HashMap.
     * * INTUITION:
     * By matching opening characters manually and pushing their respective closing
     * counterparts to the Stack, we eliminate the memory overhead and lookup cost 
     * of a traditional map.
     * * COMPLEXITY ANALYSIS:
     * - Time Complexity: O(n) where n is the length of the string.
     * - Space Complexity: O(n) worst-case memory consumption for the stack.
     */
    public boolean isValid(String s) {
        // Early exit optimization: Valid pairs require an even string length
        if (s.length() % 2 != 0) {
            return false;
        }

        Stack<Character> stack = new Stack<>();

        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);

            // Push the expected closing bracket directly
            if (ch == '(') {
                stack.push(')');
            } else if (ch == '{') {
                stack.push('}');
            } else if (ch == '[') {
                stack.push(']');
            } 
            // Validate closing bracket structure
            else {
                if (stack.isEmpty() || stack.pop() != ch) {
                    return false;
                }
            }
        }

        return stack.isEmpty();
    }
}